akka-sample-distributed-workers-scala/src/main/scala/worker/Master.scala

```diff
@@ -0,0 +1,185 @@
+package worker
+
+import java.util
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.ActorRef
+import akka.actor.typed.receptionist.{ Receptionist, ServiceKey }
+import akka.cluster.pubsub.{ DistributedPubSub, DistributedPubSubMediator }
+import akka.persistence.typed.PersistenceId
+import akka.persistence.typed.scaladsl.Effect
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import worker.WorkState.WorkAccepted
+import worker.WorkState.WorkDomainEvent
+import akka.actor.typed.scaladsl.adapter._
+import worker.WorkState.WorkCompleted
+import worker.WorkState.WorkStarted
+import worker.WorkState.WorkerFailed
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.duration.{ Deadline, FiniteDuration, _ }
+
+/**
+ * The master actor keep tracks of all available workers, and all scheduled and ongoing work items
+ */
+object Master {
+
+  val WorkerServiceKey: ServiceKey[Worker.Command] = ServiceKey[Worker.Command]("workerService")
+
+  sealed trait Command extends CborSerializable
+  // Messages from Workers
+  case class UpdatedWorkers(workers: Receptionist.Listing) extends Command
+  case class WorkerRequestsWork(workerId: String, replyTo: ActorRef[Worker.Command]) extends Command
+  case class WorkIsDone(workerId: String, workId: String, result: Any, replyTo: ActorRef[Worker.Command])
+      extends Command
+  case class WorkFailed(worker: ActorRef[Worker.Command], workId: String) extends Command
```

Comment:
I'm not sure about this. These two actors are only designed to work together. There won't be other actors that can register to receive work. In that case is it wrong to have the two actor's protocols coupled this way? What's the advantage of having separate types in the Master that are always converted to Worker messages?

Reply:
Even for component internals I think we shouldn't do circular dependencies with types, hard to follow and brittle. I agree adapting types feel like overhead, I think the types should be defined in one place so that we either have the dependency from worker to master or the other way around.

Having them with the master is probably most reasonably since it is like a protocol that the worker should support/implement to be able to work with the master?

Reply:
Okay i'll move them to the master